### ISSUE TITLE
how is 'lambda_x' updated?

### ISSUE THREAD

**Issue by LIJUNYI95** [JunjieYang97/MRVRBO#1]
I run the 'mnist_exp.py' file and it confuses me that how the outer variable 'lambda_x' is updated?  
First, only the following line of code is related to the x update: `weight = weight - args.outer_lr*outer_update`  But this won't update 'lambda_x' as 'weight' is a new variable.
Next, you create the 'lambda_x' with the size of training set but it seems that only 'spider-size' or 'batch_size' number of unique x?

**Comment by JunjieYang97** [comment]
Thanks for pointing this out. The 'lambda_x' is updated by 'weight' in the latest commit. 
In terms of sample size, stochastic algorithm updates in mini batch fashion. Hence, when we update hyperparameter ("lambda_x") every epoch, we only update part of it corresponding to sample size we set.

**Comment by LIJUNYI95** [comment]
Thanks for the reply. 
About the second point, is there a correspondence between each lambda value and sample? It looks like the samples are randomly sampled. If no correspondence, what's the interpretation for 'lambda_x'?

**Comment by JunjieYang97** [comment]
That's a very good question. 
There is no strict correspondence between lambda value and sample. The reason is that the hyperparameter gradient requires us to adopt independent samples to calculate, that’s where we introduce validation sample. In such case, the one to one corresponding relationship between lambda (hyperparameter) and training samples has been broken. Hence, the purpose of our experiments is to learn a good general "lambda_x" to be robust against poisoned data.

**Comment by LIJUNYI95** [comment]
I don't get it. Validations samples don't have noise and can be sampled from a 'clean' validation dataset (independently). While the training set is poisoned and even if each sample has a lambda, it still could be sampled independently? Can you elaborate why will the correspondence be broken?

**Comment by JunjieYang97** [comment]
For example, to calculate hyper gradient of training data "t_1", we adopt independent validation data "v_1" to calculate it. However, the real hyper gradient should be calculated by data "t_1" rather than "v_1".

**Comment by LIJUNYI95** [comment]
I see. 'B_x' needs to be independent
